File: ios/RNEwayPayments.m
#import "RNEwayPayments.h"
#import <React/RCTLog.h>

@implementation RNEwayPayments

Customer *customer;
CardDetails *cardDetails;
Payment *payment;

RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(build:(NSString *)key) {
    [[RapidAPI sharedManager] setRapidEndpoint:@"https://api.sandbox.ewaypayments.com/"];
    [[RapidAPI sharedManager] setPublicAPIKey:@"epk-2C73830F-7F82-485F-BC15-BA4BAD38F4AD"];

    customer = [[Customer alloc] init];
    cardDetails = [[CardDetails alloc] init];
    payment = [[Payment alloc] init];
}

RCT_EXPORT_METHOD(sandbox) {
    [[RapidAPI sharedManager] setRapidEndpoint:@"https://api.sandbox.ewaypayments.com/"];
}

RCT_EXPORT_METHOD(production) {
    [[RapidAPI sharedManager] setRapidEndpoint:@"https://api.ewaypayments.com/"];
}

RCT_EXPORT_METHOD(setCardDetail:(NSString *)name
                  number:(NSString *)number
                  month:(NSString *)month
                  year:(NSString *)year
                  cvn:(NSString *)cvn
                  ) {
    cardDetails.Name = name;
    cardDetails.Number = number;
    cardDetails.ExpiryMonth = month;
    cardDetails.ExpiryYear = year;
    cardDetails.CVN = cvn;
}

RCT_EXPORT_METHOD(setCustomerDetail:(NSString *)name
                  lastName:(NSString *)lastName
                  ) {
    customer.FirstName = name;
    customer.LastName = lastName;
}

RCT_EXPORT_METHOD(setPaymentDetail:(NSString *)currencyCode
                  totalAmount:(nonnull NSNumber *)totalAmount
                  invoiceDescription:(NSString *)invoiceDescription
                  ) {
    payment.Payment = totalAmount;
    payment.InvoiceDescription = invoiceDescription;
    payment.CurrencyCode = currencyCode;
}

RCT_EXPORT_METHOD(pay) {
    Transaction *transaction = [[Transaction alloc] init];
    customer.CardDetails = cardDetails;
    transaction.Customer = customer;
    transaction.Payment = payment;

    [RapidAPI submitPayment:transaction completed:^(SubmitPaymentResponse *submitPaymentResponse) {
        NSString *submissionID = submitPaymentResponse.SubmissionID;
    }];
}

@end

File: ios/RNEwayPayments.swift
//
//  RNEwayPayments.swift
//  RNEwayPayments
//
//  Created by Julian Rodriguez Nicastro on 16/07/2019.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation

// CalendarManager.swift

@objc(PaymentManager)
class PaymentManager: NSObject {
    
    private var bridge: RCTBridge!  // this is synthesized
    
    private var cardDetail : CardDetail
    private var customerDetail : CustomerDetail
    private var paymentDetail : PaymentDetail
    
    @objc func build(key: String) -> Void {
        let transaction = Transaction()

    }
    
    @objc func pay() -> Void {
        print("JUDI")
        
    }
    
    @objc func setCardDetail(name: String, number: String, month: String, year: String, cvn: String){
        self.cardDetail = CardDetail(name: name,
                                     number: number,
                                     month: month,
                                     year: year,
                                     cvn: cvn)
    }
    
    @objc func setCustomerDetail(name: String, lastName: String){
        self.customerDetail = CustomerDetail(name: name,
                                             lastName: lastName)
    }
    
    @objc func setPaymentDetail(currencyCode: String, totalAmount: Int, invoiceDescription: String){
        self.customerDetail = CustomerDetail(currencyCode: currencyCode,
                                             totalAmount: totalAmount,
                                             invoiceDescription:invoiceDescription)
    }
    
}

@objc(CardDetail)
class CardDetail: NSObject {
    
    private var name : String
    private var number : String
    private var month : String
    private var year : String
    private var cvn : String
    
    internal init(name: String, number: String, month: String, year: String, cvn: String) {
        self.name = name
        self.number = number
        self.month = month
        self.year = year
        self.cvn = cvn
    }
}

@objc(CustomerDetail)
class CustomerDetail: NSObject {
    
    private var name : String
    private var lastName : String
    
    internal init(name: String, lastName: String) {
        self.name = name
        self.lastName = number
    }
}

@objc(PaymentDetail)
class PaymentDetail: NSObject {

    private var currencyCode : String
    private var totalAmount : Int
    private var invoiceDescription : String
    
    internal init(currencyCode: String, totalAmount: Int, invoiceDescription: String) {
        self.currencyCode = currencyCode
        self.totalAmount = totalAmount
        self.invoiceDescription = invoiceDescription
    }
}
